### PR TITLE
[stm32] run optional frozen module at boot

### PR DESCRIPTION
**Pull request by fragmuffin**
Inspired by the `esp32` port:
https://github.com/micropython/micropython/blob/bbbdef4cc1d2cb3a3f6c2d067b6988534773b2a6/ports/esp32/main.c#L156

This change is optional, and can be switched on in build if 2 conditions are met:

- Your module is frozen into the firmware (ie: `manifest.py` populated, and can be imported)
- `MICROPY_BOARD_FROZEN_BOOT_FILE` switch is set to string:

For example, you create a `_boot.py` file:

```python
import pyb, time
led = led.LED(1)
for _ in range(10):
    led.toggle()
    time.sleep(0.2)
```

Add it to the `manifest.py`:

```python
freeze('_boot.py', opt=0)
```

Specify `_boot.py` in `MICROPY_BOARD_FROZEN_BOOT_FILE`

```c
#define MICROPY_BOARD_FROZEN_BOOT_FILE "_boot.py"
```

### CODE REVIEW

File: ports/stm32/main.c [hunk 1]
```diff
@@ -562,6 +562,9 @@ void stm32_main(uint32_t reset_mode) {
     MP_STATE_PORT(pyb_config_main) = MP_OBJ_NULL;
 
     // Run boot.py (or whatever else a board configures at this stage).
+    #ifdef MICROPY_BOARD_RUN_FROZEN_AT_BOOT
```
Comment (dpgeorge) [inline ports/stm32/main.c:565]: I think this should go just above the comment on line 564, with a new comment for this added code, something like "// Run optional frozen boot code."

Also, I think the config option should be `MICROPY_BOARD_FROZEN_BOOT_FILE`, to indicate that it needs to specify a filename.

Comment (thyrrestrup) [reply ports/stm32/main.c:565]: Maybe a more generic name and take state? `MICROPY_BOARD_BEFORE_BOOT(&state)`
Then the following could be done:
```c
#define MICROPY_BOARD_BEFORE_BOOT board_before_boot

void board_before_boot(boardctrl_state_t *state) {
    pyexec_frozen_module("_boot.py");
}
```

Comment (dpgeorge) [reply ports/stm32/main.c:565]: It's actually already generic the way it is (before this PR): a board can define `MICROPY_BOARD_RUN_BOOT_PY` to first execute a frozen module of its choice, then run the default `boardctrl_run_boot_py()`.

The aim of this PR is to just make it easy for a board to do the common thing of executing frozen code before the user `boot.py`.

Comment (thyrrestrup) [reply ports/stm32/main.c:565]: True!

**Review by dpgeorge — APPROVED** [review]

**Comment by dpgeorge** [comment]
Thanks for updating, merged in e83aa252f7c58db3d1a95b399c1bcdf791dcf819

**Pull request closed without merging**